File: crates/mutsuki-runtime-core/src/backend.rs
use mutsuki_runtime_contracts::{
    Envelope, LeaseToken, OperationHandlerKey, OperationSnapshot, OperationStatus, PluginSnapshot,
    RefDescriptor, ResourceRecord, SourceSnapshot, StrategyResult,
};
use serde_json::Value;

use crate::RuntimeResult;

#[derive(Clone, Debug, PartialEq)]
pub enum BackendPayload {
    Json(Value),
    Envelope(Envelope),
}

pub trait StrategyBackend {
    fn on_awake(&mut self, agent_id: &str) -> RuntimeResult<()>;
    fn on_input(&mut self, agent_id: &str, envelope: &Envelope) -> RuntimeResult<StrategyResult>;
    fn next_step(&mut self, agent_id: &str) -> RuntimeResult<StrategyResult>;
    fn on_stop(&mut self, agent_id: &str) -> RuntimeResult<()>;
}

pub trait OperationBackend {
    fn list_plugins(&self) -> RuntimeResult<Vec<PluginSnapshot>>;
    fn list_operations(
        &self,
        enabled_plugin_ids: &[String],
    ) -> RuntimeResult<Vec<OperationSnapshot>>;
    fn list_sources(&self, enabled_plugin_ids: &[String]) -> RuntimeResult<Vec<SourceSnapshot>>;
    fn invoke(
        &mut self,
        agent_id: &str,
        key: &OperationHandlerKey,
        payload: Value,
    ) -> RuntimeResult<BackendPayload>;
    fn operation_status(&self, agent_id: &str, key: &OperationHandlerKey) -> OperationStatus;
}

pub trait ResourceBackend {
    fn register_resource(
        &mut self,
        descriptor: RefDescriptor,
        owner: &str,
    ) -> RuntimeResult<String>;
    fn acquire_resource(&mut self, ref_id: &str, requester: &str) -> RuntimeResult<LeaseToken>;
    fn release_resource(&mut self, token: &LeaseToken) -> RuntimeResult<()>;
    fn list_records(&self, owner: Option<&str>) -> Vec<ResourceRecord>;
}

pub trait RuntimeBackend: StrategyBackend + OperationBackend {}

impl<T> RuntimeBackend for T where T: StrategyBackend + OperationBackend {}

File: python/mutsukicore-runtime-python/src/mutsukicore_runtime_python/backend.py
from __future__ import annotations

from typing import Protocol

from mutsukicore_runtime_python.contracts import (
    Envelope,
    JsonValue,
    LeaseToken,
    OperationHandlerKey,
    OperationSnapshot,
    OperationStatus,
    PluginSnapshot,
    RefDescriptor,
    ResourceRecord,
    RuntimeError,
    SourceSnapshot,
    StrategyResult,
)


class BackendInvokeError(Exception):
    """Structured backend failure wrapper."""

    def __init__(self, error: RuntimeError) -> None:
        super().__init__(f"runtime backend failed: {error.code}")
        self.error = error


class StrategyBackend(Protocol):
    async def on_awake(self, agent_id: str) -> None: ...

    async def on_input(self, agent_id: str, envelope: Envelope) -> StrategyResult: ...

    async def next_step(self, agent_id: str) -> StrategyResult: ...

    async def on_stop(self, agent_id: str) -> None: ...


class OperationBackend(Protocol):
    def list_plugins(self) -> tuple[PluginSnapshot, ...]: ...

    def list_operations(
        self, enabled_plugin_ids: tuple[str, ...] | list[str]
    ) -> tuple[OperationSnapshot, ...]: ...

    def list_sources(
        self, enabled_plugin_ids: tuple[str, ...] | list[str]
    ) -> tuple[SourceSnapshot, ...]: ...

    async def invoke(
        self,
        agent_id: str,
        key: OperationHandlerKey,
        payload: JsonValue = None,
    ) -> JsonValue: ...

    def operation_status(self, agent_id: str, key: OperationHandlerKey) -> OperationStatus: ...


class ResourceBackend(Protocol):
    async def register_resource(self, descriptor: RefDescriptor, owner: str) -> str: ...

    async def acquire_resource(self, ref_id: str, requester: str) -> LeaseToken: ...

    async def release_resource(self, token: LeaseToken) -> None: ...

    def list_records(self, owner: str | None = None) -> tuple[ResourceRecord, ...]: ...
